File: localdrive/android/app/src/main/kotlin/app/localdrive/PresenceBeacon.kt
package app.localdrive

import android.content.Context
import android.net.nsd.NsdManager
import android.net.nsd.NsdServiceInfo
import android.net.wifi.WifiManager

/**
 * "This person, on this network, right now."
 *
 * Advertised over mDNS while a sharing screen is open, and never otherwise.
 * Broadcasting your presence at all times is not something an app should do
 * without asking, and it would cost battery for nothing the rest of the time.
 *
 * This changes discovery and presentation only. The share itself is still a
 * permission grant against the server, so the person on the other end gets an
 * ongoing shared item they can open next week from a different device, not a
 * one-off copy beamed between two phones.
 */
class PresenceBeacon(private val context: Context) {

    private var nsdManager: NsdManager? = null
    private var listener: NsdManager.RegistrationListener? = null
    private var multicastLock: WifiManager.MulticastLock? = null

    fun start(displayName: String, userId: String, avatarSeed: String) {
        if (listener != null) stop()

        // Android drops multicast packets by default to save power, so
        // discovering anyone else requires holding this open for the duration
        val wifi = context.applicationContext
            .getSystemService(Context.WIFI_SERVICE) as WifiManager
        multicastLock = wifi.createMulticastLock(LOCK_NAME).apply {
            setReferenceCounted(true)
            acquire()
        }

        val info = NsdServiceInfo().apply {
            // the instance name is what other devices see before they resolve
            // the record, so it carries nothing private
            serviceName = "ld-${userId.take(8)}"
            serviceType = SERVICE_TYPE
            port = PRESENCE_PORT
            setAttribute("name", displayName)
            setAttribute("uid", userId)
            setAttribute("seed", avatarSeed)
        }

        val registration = object : NsdManager.RegistrationListener {
            override fun onServiceRegistered(info: NsdServiceInfo) = Unit
            override fun onRegistrationFailed(info: NsdServiceInfo, code: Int) = Unit
            override fun onServiceUnregistered(info: NsdServiceInfo) = Unit
            override fun onUnregistrationFailed(info: NsdServiceInfo, code: Int) = Unit
        }
        listener = registration

        nsdManager = (context.getSystemService(Context.NSD_SERVICE) as NsdManager).also {
            it.registerService(info, NsdManager.PROTOCOL_DNS_SD, registration)
        }
    }

    fun stop() {
        listener?.let { current ->
            // unregistering can throw if registration never completed, which
            // is not worth crashing over when the screen is already closing
            runCatching { nsdManager?.unregisterService(current) }
        }
        listener = null
        nsdManager = null

        runCatching { multicastLock?.release() }
        multicastLock = null
    }

    companion object {
        const val SERVICE_TYPE = "_ldpeer._tcp"

        /**
         * Nothing listens on it. mDNS requires a port in the SRV record, and
         * the beacon is presence only: the actual share goes through the
         * server, so there is no socket for anyone to connect to here.
         */
        private const val PRESENCE_PORT = 7444

        private const val LOCK_NAME = "localdrive-presence"
    }
}

File: localdrive/ios/Runner/PresenceBeacon.swift
import Foundation

/// "This person, on this network, right now."
///
/// Advertised over Bonjour while a sharing screen is open, and never
/// otherwise. Broadcasting presence at all times is not something an app should
/// do without asking, and it costs battery for nothing the rest of the time.
///
/// This changes discovery and presentation only. The share itself is still a
/// permission grant against the server, so the person on the other end gets an
/// ongoing shared item they can open next week from a different device, rather
/// than a one-off copy beamed between two phones.
final class PresenceBeacon: NSObject, NetServiceDelegate {

    static let serviceType = "_ldpeer._tcp"

    /// Nothing listens on it. Bonjour requires a port, and this beacon is
    /// presence only: the actual share goes through the server, so there is no
    /// socket for anyone to connect to here.
    private static let presencePort: Int32 = 7444

    private var service: NetService?

    func start(displayName: String, userId: String, avatarSeed: String) {
        stop()

        // the instance name is what other devices see before they resolve the
        // record, so it carries nothing private
        let instance = "ld-" + String(userId.prefix(8))
        let service = NetService(
            domain: "local.",
            type: Self.serviceType,
            name: instance,
            port: Self.presencePort
        )

        let record: [String: Data] = [
            "name": Data(displayName.utf8),
            "uid": Data(userId.utf8),
            "seed": Data(avatarSeed.utf8),
        ]
        service.setTXTRecordData(NetService.data(fromTXTRecord: record))
        service.delegate = self
        service.publish()

        self.service = service
    }

    func stop() {
        service?.stop()
        service?.delegate = nil
        service = nil
    }

    func netService(_ sender: NetService, didNotPublish errorDict: [String: NSNumber]) {
        // usually the Local Network permission being declined. Nothing to
        // recover: the picker falls back to its plain list, which works
        stop()
    }
}
